### PR TITLE
Improve in_array ConstantArrayType handling

### PR DESCRIPTION
**Pull request by herndlm**
well, another try of improving in_array, this time ConstantArrayType-specific only. It seems to be adding value, but I don't like that one conditional and that ConstantArrayType need this extra handling. I didn't find a way around both yet though.

### CODE REVIEW

File: src/Type/Php/InArrayFunctionTypeSpecifyingExtension.php [hunk 1]
```diff
@@ -85,6 +101,51 @@ public function specifyTypes(FunctionReflection $functionReflection, FuncCall $n
...
+				static function (Type $offsetType, callable $traverse) use ($context, $arrayType, $needleType): Type {
+					if ($offsetType instanceof UnionType || $offsetType instanceof IntersectionType) {
+						return $traverse($offsetType);
+					}
+
+					$resultArray = $arrayType;
+					if ($context->truthy()) {
+						$resultArray = $resultArray->makeOffsetRequired($offsetType);
+					} elseif ($offsetType instanceof ConstantIntegerType && $resultArray->isOptionalKey($offsetType->getValue())) {
+						$resultArray = $resultArray->unsetOffset($offsetType);
```
Comment (herndlm) [inline src/Type/Php/InArrayFunctionTypeSpecifyingExtension.php:116]: it should be able to handle more types here now I suppose

File: src/Type/Constant/ConstantArrayType.php [hunk 2]
```diff
@@ -564,6 +564,39 @@ public function getOffsetValueType(Type $offsetType): Type
...
+			if ($type instanceof ErrorType) {
+				return new MixedType();
+			}
+
+			return $type;
+		}
+
+		return new ErrorType(); // undefined offset value
+	}
+
```
Comment (herndlm) [inline src/Type/Constant/ConstantArrayType.php:599]: this is very much inspired (= copy/pasted and replaced keys and offsets xD) by `getOffsetValueType`

**Comment by herndlm** [comment]
This is still bad, e.g. what about unions of constant arrays.. But at least no unexpected CI errors

**Comment by ondrejmirtes** [comment]
Hi, I'm cleaning up old and stale PRs. Please send a new PR if you're still interested, thanks.

We should probably revisit this later once we learn a bit more about the problem domain :)

**Pull request closed without merging**